rules/aip0191/csharp_namespace.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0191
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/locations"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/stoewer/go-strcase"
+)
+
+var csharpNamespace = &lint.FileRule{
+	Name: lint.NewRuleName(191, "csharp-namespace"),
+	LintFile: func(f *desc.FileDescriptor) []lint.Problem {
+		if ns := f.GetFileOptions().GetCsharpNamespace(); ns != "" {
+			// Check for invalid characters.
+			if !csharpValidChars.MatchString(ns) {
+				return []lint.Problem{{
+					Message:    "Invalid characters: C# namespaces only allow [A-Za-z0-9.].",
+					Descriptor: f,
+					Location:   locations.FileCsharpNamespace(f),
+				}}
+			}
+
+			// Check that upper camel case is used.
+			upperCamel := []string{}
+			for _, segment := range strings.Split(ns, ".") {
+				wantSegment := csharpVersionRegexp.ReplaceAllStringFunc(
+					strcase.UpperCamelCase(segment),
```

Comment:
Will this be okay with Google.Cloud.AutoML.V1 for example?

Reply:
Discussed in video chat. Sadly no, it wants `AutoMl`. Most likely we merge this as-is and they have to disable it, because I can not think of a good way around it.